### PR TITLE
Fix PluginManager state machine

### PR DESCRIPTION
**Pull request by fmitchell-r7**
This replaces the bootstrap logic in PluginManager by listening for updates and retrying on a timer. The end result is slightly slower boot time in the worst case. If data from the EC2 metadata or S3 index are unavailable, the PluginManager waits until it gets an update form one or the other before parsing the index for additional sources.

### CODE REVIEW

**Comment by dgreene-r7** [comment]
:+1:

**Comment by fmitchell-r7** [comment]
Waiting until @dgreene-r7 has the plugin cleanup logic worked out before merging this. We want to see how the two interact.

File: lib/storage.js [hunk 1]
```diff
@@ -88,6 +89,10 @@ class Storage {
    * @param {Object} plugin
    */
   unregister(plugin) {
+    // TODO: WRITE TESTS
+    if (typeof plugin.shutdown === typeof (Function)) {
```
Comment (jmanero-r7) [inline lib/storage.js]: Use `plugin.shutdown instanceof Function` instead.

**Comment by jmanero-r7** [comment]
:ship: :rocket: :banana:

**Pull request merged**